### PR TITLE
Added generic type parameter overload to ModelSet create method

### PR DESCRIPTION
**Pull request by StuporHero**
Thank you very much for your contribution. We highly appreciate the time
and effort you put into pull request. Before you submit the pull request,
please review the [Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md#contribution-workflow).

Make sure you check the following boxes.

For all changes:

- [x] Sign the [Gradle CLA](http://gradle.org/contributor-license-agreement/).

For all non-trivial changes that modify the behavior or public API:

- [x] Before submitting a pull request, you started a discussion on the
[Gradle developer list](https://groups.google.com/forum/#!forum/gradle-dev)
or can reference a [JIRA issue](https://issues.gradle.org/secure/Dashboard.jspa).

https://groups.google.com/forum/#!topic/gradle-dev/GLZkTEE_lZk

- [x] You considered writing a design document. A design document can be
brief but explains the use case or problem you are trying to solve,
touches on the planned implementation approach as well as the test cases
that verify the behavior. Optimally, design documents should be submitted
as a separate pull request. You can find [samples](https://github.com/gradle/gradle/tree/master/design-docs)
in the Gradle GitHub repository. Please let us know if you need help with
creating the design document. We are happy to help!
- [x] The pull request contains an appropriate level of unit and integration
test coverage to verify the behavior. Before submitting the pull request
you ran a build on your local machine via the command `./gradlew quickCheck`
and the tests for the modified module.
- [x] The pull request updates the Gradle documentation like user guide,
DSL reference and Javadocs where applicable.

### CODE REVIEW

File: subprojects/model-core/src/main/java/org/gradle/model/internal/core/NodeBackedModelSet.java [hunk 1]
```diff
@@ -71,17 +71,22 @@ public String toString() {
...
+    public void create(Action<? super T> action) {
+        create(getElementTypeClass(), action);
+    }
+
+    @Override
+    public <S extends T> void create(Class<S> type, final Action<? super T> action) {
         state.assertCanMutate();
 
-        String name = String.valueOf(modelNode.getLinkCount(ModelNodes.withType(elementType)));
+        String name = String.valueOf(modelNode.getLinkCount(ModelNodes.withType(type)));
```
Comment (eskatos) [inline subprojects/model-core/src/main/java/org/gradle/model/internal/core/NodeBackedModelSet.java]: This won't work when mixing different types in a `ModelSet` as node paths will overlap.
Simply reverting the change on that line fix the issue.

File: subprojects/model-core/src/test/groovy/org/gradle/model/internal/manage/projection/ModelSetModelProjectionTest.groovy [hunk 2]
```diff
@@ -73,6 +76,20 @@ class ModelSetModelProjectionTest extends ProjectRegistrySpec {
...
+            create(NamedSubThing) { name = '2' }
+        }
+
+        then:
+        def set = registry.realize(collectionPath, collectionType)
+        set*.name == ['1', '2']
+        set.toArray().collect { it.name } == ['1', '2']
+        set.toArray(new NamedSubThing[2]).collect { it.name } == ['1', '2']
+    }
+
```
Comment (eskatos) [inline subprojects/model-core/src/test/groovy/org/gradle/model/internal/manage/projection/ModelSetModelProjectionTest.groovy]: Could you please add some coverage for when elements of different types are mixed in the same `ModelSet`?

File: subprojects/model-core/src/main/java/org/gradle/model/internal/core/NodeBackedModelSet.java [hunk 3]
```diff
@@ -165,9 +170,39 @@ public void clear() {
...
+    }
+
+    public void create(Class<? extends T> type, Closure<?> closure) {
+        if (hasNoParameters(closure)) {
+            create(ClosureBackedAction.of(closure));
+        } else if (hasOneParameter(closure)) {
+            validateClosureParameter(type, closure);
+            create(type, ClosureBackedAction.of(closure));
+        } else {
+            throw new TooManyClosureParametersException(type);
```
Comment (eskatos) [inline subprojects/model-core/src/main/java/org/gradle/model/internal/core/NodeBackedModelSet.java]: This should throw an `IllegalArgumentException` instead to be on par with what `ModelMap` does, see [NodeBackedModelMap.java#L528](https://github.com/gradle/gradle/blob/f490bdf61bd9b4f5383cd9fb0d8ffbca93da8c32/subprojects/model-core/src/main/java/org/gradle/model/internal/core/NodeBackedModelMap.java#L528)

File: subprojects/model-core/src/main/java/org/gradle/model/internal/core/NodeBackedModelSet.java [hunk 4]
```diff
@@ -165,9 +170,39 @@ public void clear() {
...
+        return closure.getParameterTypes().length == 0;
+    }
+
+    private boolean hasOneParameter(Closure<?> closure) {
+        return closure.getParameterTypes().length == 1;
+    }
+
+    private void validateClosureParameter(Class<? extends T> type, Closure<?> closure) {
+        if (!Object.class.equals(closure.getParameterTypes()[0]) && !type.equals(closure.getParameterTypes()[0])) {
+            throw new InvalidModelSetClosureParameterException(type, closure);
```
Comment (eskatos) [inline subprojects/model-core/src/main/java/org/gradle/model/internal/core/NodeBackedModelSet.java]: Same thing here.

**Comment by eskatos** [comment]
Hi Michael, thanks for your contribution!

I've added some comments to your changes.
Your effort on documentation is truly appreciated.

**Comment by StuporHero** [comment]
Thank you very much! I definitely needed the feedback. I'll turn that around a.s.a.p.

File: subprojects/model-core/src/test/groovy/org/gradle/model/internal/manage/projection/ModelSetModelProjectionTest.groovy (hunk 2 above)
Comment (StuporHero) [inline subprojects/model-core/src/test/groovy/org/gradle/model/internal/manage/projection/ModelSetModelProjectionTest.groovy]: Should mixed types be a separate test, or can I just modify that one? Obviously, closure parameter validation is a separate test.

**Comment by StuporHero** [comment]
Ready for your review.

File: subprojects/model-core/src/main/java/org/gradle/model/internal/core/NodeBackedModelSet.java (hunk 1 above)
Comment (StuporHero) [inline subprojects/model-core/src/main/java/org/gradle/model/internal/core/NodeBackedModelSet.java]: I guess node paths aren't entirely clear to me yet. Which classes should I look at to get a better grasp of paths? I'd be happy to spruce up the documentation if I can grok the subject.

Comment (eskatos) [inline subprojects/model-core/src/main/java/org/gradle/model/internal/core/NodeBackedModelSet.java]: See the [Model paths](https://docs.gradle.org/current/userguide/software_model.html#N179BB) section of the user guide for a quick but accurate explanation.

In this particular case, we generate `ModelSet` elements paths by appending their index. So a `ModelSet` at `things` path will have elements at `things.0`, `things.1` etc.. That's why we need to count the elements with the base element type.

File: subprojects/model-core/src/main/java/org/gradle/model/ModelSet.java [hunk 5]
```diff
@@ -18,6 +18,7 @@
 
 import org.gradle.api.Action;
 import org.gradle.api.Incubating;
+import org.gradle.model.internal.type.ModelType;
```
Comment (eskatos) [inline subprojects/model-core/src/main/java/org/gradle/model/ModelSet.java]: This import is not used.

File: subprojects/model-core/src/main/java/org/gradle/model/internal/core/NodeBackedModelSet.java [hunk 6]
```diff
@@ -28,6 +28,7 @@
 import org.gradle.model.internal.manage.instance.ManagedInstance;
 import org.gradle.model.internal.type.ModelType;
 
+import java.lang.reflect.Type;
```
Comment (eskatos) [inline subprojects/model-core/src/main/java/org/gradle/model/internal/core/NodeBackedModelSet.java]: This import is not used.

**Comment by eskatos** [comment]
Thanks for the changes. I added a couple more minor comments.
Otherwise, the PR is now in a good shape.
I can do the squashing myself but feel free to rework the PR history if you feel like so.

**Comment by eskatos** [comment]
Hi Michael,

I made an attempt at merging this but it looks like I didn't run enough checks before that.
Some integration tests are still failing.
I reverted the merge in 840b586756184528843998ee709fe56331c352cd.

The two offending tests are: `ModelSetIntegrationTest` and `UserGuideSamplesIntegrationTest`.

In `ModelSetIntegrationTest`, model creation of "ModelSet of ModelSets" and "ModelSet of ModelMaps" seems to break, could you take a look?

`UserGuideSamplesIntegrationTest` fails because of small mistakes in `subprojects/docs/src/samples/modelRules/modelDsl/build.gradle`.
You can run it locally using the following incantation:
`./gradlew intTestImage docs:userguide integtest:integTest --tests org.gradle.integtests.samples.UserGuideSamplesIntegrationTest`

So, given that we cannot "reopen" a PR, could you please open another one so we can push this forward?

**Comment by eskatos** [comment]
I see that you just dropped your branch.
For your convenience this will bring back all the changes from this PR into your working copy:
`git cherry-pick --no-commit 0d39dc353a702adc9ed070b491a720ce9805e6de`

**Comment by StuporHero** [comment]
I'll recover that branch and create a new PR tomorrow.

**Comment by StuporHero** [comment]
PR #637 is the new PR.

**Pull request merged**